### PR TITLE
(docs) update documentation wrt functions moved to puppet

### PR DESCRIPTION
**Pull request by hlindberg**

### CODE REVIEW

**Comment by hlindberg** [comment]
Caught by a style checker that does not like an extra space... 
image

**Comment by pmcmaw** [comment]
Ran through our internal adhoc pipeline with no issues and all changes look as expected.
Happy enough when @clairecadman has checked the docs to merge.

**Comment by hlindberg** [comment]
@pmcmaw I just pushed one more commit (after your check). No more commits coming now - promise... (unless someone wants something changed).

**Comment by pmcmaw** [comment]
So running this through the adhoc pipeline doe not resolve the issue. After some high level looking it seems to be continuing to run the test. I am carrying out some investigation into this.

**Comment by pmcmaw** [comment]
Hey @hlindberg , 

So after investigation and confirmation on our internal adhoc pipelines would it be possible to replace ```Puppet.version`` with ```return_puppet_version`` ` which is a function that already exists in the spec_helper_acceptance.rb. This should resolve the failures and is only required in the acceptance tests.

Here is an example that I have done for size.rb: #925 

Apologies!

**Comment by hlindberg** [comment]
@pmcmaw PR updated, I changed all of the acceptance tests for functions checking with Puppet.version and 6.0.0

File: lib/puppet/parser/functions/concat.rb [hunk 1]
```diff
@@ -12,6 +12,10 @@ module Puppet::Parser::Functions
     Would result in:
 
       ['1','2','3','4','5','6','7','8','9']
+
+    Note: Since puppet 4.0 concatenation of arrays and hashes can be done with the + operator.
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/concat.rb]: The 'p' in Puppet should be uppercase.

**Comment by hlindberg** [comment]
@pmcmaw @clairecadman All requested changes are now in this PR. LGTM

File: lib/puppet/parser/functions/count.rb [hunk 2]
```diff
@@ -4,8 +4,20 @@
...
     Takes an array as first argument and an optional second argument.
-    Count the number of elements in array that matches second argument.
-    If called with only an array it counts the number of elements that are not nil/undef.
+    Count the number of elements in array that is equal to the second argument.
+    If called with only an array it counts the number of elements that are not nil/undef/empty-string.
+
+    Note: equality is tested with a Ruby method and it is therefore subject to what Ruby considers
+    to be equal. For strings this means that equality is case sensitive.
+
+    In Puppet core, counting can be done in general by using a combination of the core functions
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/count.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/delete.rb [hunk 3]
```diff
@@ -19,6 +19,22 @@ module Puppet::Parser::Functions
 
         delete('abracadabra', 'bra')
         Would return: 'acada'
+
+    Note that from puppet 4.0.0 the minus (-) operator deletes values from arrays and keys from a hash:
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/delete.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/delete_undef_values.rb [hunk 4]
```diff
@@ -15,6 +15,11 @@ module Puppet::Parser::Functions
 
     Would return: ['A','',false]
 
+    Note that since Puppet 4.0.0 the equivalent can be performed with the filter() function in puppet:
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/delete_undef_values.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/delete_values.rb [hunk 5]
```diff
@@ -11,6 +11,11 @@ module Puppet::Parser::Functions
 
     Would return: {'a'=>'A','c'=>'C','B'=>'D'}
 
+    Note that since Puppet 4.0.0 the equivalent can be performed with the filter() function in puppet:
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/delete_values.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/difference.rb [hunk 6]
```diff
@@ -12,6 +12,12 @@ module Puppet::Parser::Functions
         difference(["a","b","c"],["b","c","d"])
 
     Would return: ["a"]
+
+    Note: Since Puppet 4 the minus (-) operator in the puppet language does the same thing:
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/difference.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/get_module_path.rb [hunk 7]
```diff
@@ -8,6 +8,9 @@ module Puppet::Parser::Functions
 
     Example:
       $module_path = get_module_path('stdlib')
+
+    Note that since Puppet 5.4.0 the function `module_directory()` in puppet does the same thing and will return
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/get_module_path.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/getparam.rb [hunk 8]
```diff
@@ -5,20 +5,38 @@
...
+          # This will notice the value of the parameter
+          notice(getparam(Example_resource["example_resource_instance"], "param"))
+        }
+
+        # Declare an instance of the second resource type - this will call notice
+        example_get_param { 'show_notify': }
+
+    Would notice: 'the value we are getting in this example'
+
+    Note that since puppet 4.0.0 it is possible to get a parameter value by using its data type
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/getparam.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/grep.rb [hunk 9]
```diff
@@ -13,6 +13,10 @@ module Puppet::Parser::Functions
     Would return:
 
         ['aaa','aaaddd']
+
+    Note that since Puppet 4.0.0, the filter() function in puppet can do the same:
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/grep.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/count.rb (hunk 2 above)
Comment (hlindberg) [reply lib/puppet/parser/functions/count.rb]: Hm, the line in question has `Puppet` with capital `P` already ???

File: lib/puppet/parser/functions/has_key.rb [hunk 10]
```diff
@@ -15,6 +15,12 @@ module Puppet::Parser::Functions
           notice('this will be printed')
         }
 
+    Note: Since Puppet 4.0.0 this can be achieved in the puppet language with the following equivalent expression:
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/has_key.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/hash.rb [hunk 11]
```diff
@@ -10,6 +10,13 @@ module Puppet::Parser::Functions
         hash(['a',1,'b',2,'c',3])
 
     Would return: {'a'=>1,'b'=>2,'c'=>3}
+
+    Note: Since Puppet 5.0.0 type conversions can in general be performed by using the Puppet Type System.
+    See the function new() in puppet for a wide range of available type conversions.
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/hash.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/hash.rb [hunk 12]
```diff
@@ -10,6 +10,13 @@ module Puppet::Parser::Functions
...
+    This example shows the equivalent expression in the puppet language:
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/hash.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/member.rb [hunk 13]
```diff
@@ -25,6 +25,19 @@ module Puppet::Parser::Functions
         member(['a', 'b', 'c'], ['d', 'b'])
 
     would return: false
+
+    Note: Since Puppet 4.0.0 the same can be performed in the puppet language. For single values
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/member.rb]: The 'p' in Puppet should be uppercase.

File: lib/puppet/parser/functions/prefix.rb [hunk 14]
```diff
@@ -10,6 +10,11 @@ module Puppet::Parser::Functions
         prefix(['a','b','c'], 'p')
 
     Will return: ['pa','pb','pc']
+
+    Note that since Puppet 4.0.0 the general way to modify values is in array is by using the map
```
Comment (pmcmaw) [inline lib/puppet/parser/functions/prefix.rb]: the line break doesn't need to be there and The 'p' in Puppet should be uppercase.

**Comment by pmcmaw** [comment]
@clairecadman if you approve these changes I will get this merged 👍

**Comment by hlindberg** [comment]
There are still the lower case 'puppet' in function files to address (I did not get that done yesterday, only fixed README.md). Merge and fix later, or fix now?

**Comment by pmcmaw** [comment]
Ill do it later, would be nice to get this in 👍
Squash and merge isnt enabled for this repo (apparently?) So going to try to enable it then will get it in.

**Comment by clairecadman** [comment]
I can go back in a fix these later today! But feel free to merge first if
you want to get it in.

**Comment by hlindberg** [comment]
@pmcmaw The many commits are done on purpose as it makes it easier to later get relevant information when looking at "blame" for some lines (they have link to the puppet tickets that caused the doc change for example). A squash will give you a commit message containing all of the comments I made about each change for those lines. Anyway, I don't object to squashing, just my 2c for why I did many commits.

**Pull request merged**